staging/src/k8s.io/component-base/flagz/flagz.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flagz
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/spf13/pflag"
+	"k8s.io/apiserver/pkg/endpoints/metrics"
+
+	cliflag "k8s.io/component-base/cli/flag"
+)
+
+type Registry struct {
+	path           string
+	lock           sync.Mutex
+	flags          []*cliflag.NamedFlagSets
+	flagsInstalled bool
+}
+
+// Flagz installs the flagz handler
+type Flagz struct{}
+
+type mux interface {
+	Handle(path string, handler http.Handler)
+}
+
+// Install adds the DefaultFlagz handler
+func (f Flagz) Install(m mux, flags []*cliflag.NamedFlagSets) {
+	flagzRegistry := Registry{
+		flags: flags,
+	}
+	flagzRegistry.installHandler(m)
+}
+
+func (reg *Registry) installHandler(m mux) {
+	reg.lock.Lock()
+	defer reg.lock.Unlock()
+	reg.flagsInstalled = true
+	m.Handle("/flagz",
+		metrics.InstrumentHandlerFunc("GET",
+			/* group = */ "",
+			/* version = */ "",
+			/* resource = */ "",
+			/* subresource = */ "/flagz",
+			/* scope = */ "",
+			/* component = */ "",
+			/* deprecated */ false,
+			/* removedRelease */ "",
+			handleFlags(reg.flags)))
+}
+
+// handleFlags returns an http.HandlerFunc that serves the provided flags.
+func handleFlags(flags []*cliflag.NamedFlagSets) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var individualCheckOutput bytes.Buffer
```

Comment:
We should be able to build the output only once when installing the handler since the flags can't really be set dynamically no? If the flags are modified, kubelet will recreate the kube-apiserver static pod.

I was wondering if `/debug/flags/v` could invalidate that, but it seems like it is not modifying the flags at all, probably to avoid triggering kubelet:
https://github.com/kubernetes/kubernetes/blob/e386aa0ac4b8cb07a68bd482f11c77698e7691fa/staging/src/k8s.io/apiserver/pkg/server/config.go#L1085
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/logs/logs.go#L187-L209
But it might be worth verifying as I could've missed a layer

Reply:
Yeah I think you're right, flags cant be changed dynamically. I updated the code now to only generate the response once and reuse the same response for subsequent accesses of the /flagz endpoint